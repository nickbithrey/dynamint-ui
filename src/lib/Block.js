import React from 'react';
import { Stack } from 'office-ui-fabric-react/lib/Stack';

const Block = ({children, ...remainingProps}) => (
	<Stack.Item {...remainingProps}>
		{children}
	</Stack.Item>
);

export default Block;