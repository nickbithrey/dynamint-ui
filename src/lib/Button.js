import React from 'react';
import PropTypes from 'prop-types';
import { 
	PrimaryButton, 
	DefaultButton, 
	CommandBarButton, 
	IconButton 
} from 'office-ui-fabric-react/lib/Button';

const buttonTypes = {
	standard: DefaultButton,
	primary: PrimaryButton,
	command: CommandBarButton,
	icon: IconButton
};

const Button = ({text, onClick, type}) => {
	const Btn = buttonTypes[type];
	return <Btn text={text} onClick={onClick} />
};

Button.defaultProps = {
	type: 'standard'
};

Button.propTypes = {
	text: PropTypes.string.isRequired,
	onClick: PropTypes.func
}

export default Button;

export const newBtn = (text, onClick, condition = () => true) => (
	{
		text,
		onClick,
		condition,
		type: 'standard'
	}
);

export const newPrimaryBtn = (text, onClick, condition = () => true) => (
	{
		text,
		onClick,
		condition,
		type: 'primary'
	}
);