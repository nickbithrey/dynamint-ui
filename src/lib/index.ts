export interface Key {
    key: string | number;
}

export interface Condition<T> {
    condition: (input: T) => boolean;
}

export interface Loading {
    loading: boolean;
}

export interface Items<T> {
    items: Array<T>;
}

export interface URI {
    uri: string;
}

export interface Display {
    readOnly: boolean;
}

export interface IURILink {
    href: string;
}