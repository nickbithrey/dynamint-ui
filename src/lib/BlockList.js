import React from 'react';
import PropTypes from 'prop-types';
import { List } from 'office-ui-fabric-react/lib/List';

const renderCell = (tag, updateEntry) => {
	const CellTag = tag;
	return (item, index) => {
		return <CellTag item={item} index={index} update={updateEntry} />;
	};
};

const BlockList = ({items, tag, updateEntry}) => {
	return (
		<List
	        items={items}
	        onRenderCell={renderCell(tag, updateEntry)}
	    />
	);
};

BlockList.propTypes = {
	items: PropTypes.arrayOf(PropTypes.object).isRequired,
	tag: PropTypes.func.isRequired
}

export default BlockList;