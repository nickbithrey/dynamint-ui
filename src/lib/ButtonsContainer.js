import React from 'react';
import Btn from '~/lib/Button';
import { Columnar, Block } from '~/lib/Grid';

const ButtonsContainer = ({children = [], buttons = []}) => (
	<Columnar horizontalAlign={'end'} tokens={{childrenGap: 5}}>
		{renderChildren(children)}
		{buttons.map(renderButton)}
	</Columnar>
);

const renderChildren = (children) => {
	if (Array.isArray(children)) {
		return children.map(renderChild)
	} else if (children) {
		return renderChild(children, 0);
	}
}

const renderChild = (child, i) => (
	<Block key={i}>
		{child}
	</Block>
);

const renderButton = (button, i) => {
	if (!button.condition()) {
		return;
	}
	return (
		<Block key={i}>
			<Btn {...button} />
		</Block>
	);
}

export default ButtonsContainer;