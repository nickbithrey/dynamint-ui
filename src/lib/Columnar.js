import React from 'react';
import { Stack } from 'office-ui-fabric-react/lib/Stack';

const Columnar = ({children, ...remainingProps}) => (
	<Stack horizontal horizontalAlign="space-between" {...remainingProps}>
		{children}
	</Stack>
);

export default Columnar;