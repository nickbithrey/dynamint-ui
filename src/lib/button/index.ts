import { Button, IButton } from './Button';
import { ButtonContainer, IButtonContainer } from './ButtonContainer';

export { 
    Button,
    IButton,
    ButtonContainer,
    IButtonContainer
};