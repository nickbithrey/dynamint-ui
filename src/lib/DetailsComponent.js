import React from 'react';
import PropTypes from 'prop-types';
import { newBtn, newPrimaryBtn } from '~/lib/Button';
import Loading from '~/lib/Loading';
import { Rowise, Block } from '~/lib/Grid';
import ButtonsContainer from '~/lib/ButtonsContainer';

const detailsComponent = (WrappedComponent) => {
	class DetailsComponent extends React.Component {
		
		constructor(props) {
			super(props);
			this.load = this.load.bind(this);
			this.renderTitle = this.renderTitle.bind(this);
			this.createBackBtn = this.createBackBtn.bind(this);
			this.createSaveBtn = this.createSaveBtn.bind(this);
			this.back = this.back.bind(this);
			this.handleUpdateField = this.handleUpdateField.bind(this);
			this.handleUpdate = this.handleUpdate.bind(this);
			this.uri = props.location.state.uri;
			this.isCreate = props.location.state.isCreate;
		}
		
		load() {
			this.props.load(this.uri, this.isCreate);
		}
		
		componentDidMount() {
			this.load();
		}
		
		componentWillUnmount() {
			this.props.clear();
		}
		
		handleUpdateField(field) {
			this.props.updateField(field);
		}
		
		handleUpdate(e) {
			e.preventDefault();
			this.props.update(this.props.record, this.uri, this.isCreate)
			this.back();
		}
		
		createSaveBtn() {
			return newPrimaryBtn('Save', this.handleUpdate);
		}
		
		render() {
			if (this.props.loading || !this.props.record) {
				return <Loading />
			}
			const {
				updateField,
				update,
				...remainingProps
			} = this.props;
			return (
				<div>
					{this.renderTitle()}
					<Rowise tokens={{childrenGap: 5}}>
						<Block>
							<WrappedComponent fieldUpdate={this.handleUpdateField} update={this.handleUpdate} {...remainingProps} />
						</Block>
						<Block>
							<ButtonsContainer buttons={[this.createSaveBtn(), this.createBackBtn()]} />
						</Block>
					</Rowise>
				</div>
			);
		}
		
		renderTitle() {
			const {
				title
			} = this.props;
			return <h1>{title}</h1>;
		}
		
		back() {
			const {
				history
			} = this.props;
			history.goBack();
		}
		
		createBackBtn() {
			return newBtn('Back', this.back);
		}
		
	}

	DetailsComponent.propTypes = {
		load: PropTypes.func.isRequired,
		clear: PropTypes.func.isRequired,
		update: PropTypes.func.isRequired,
		updateField: PropTypes.func.isRequired,
		loading: PropTypes.bool.isRequired,
		history: PropTypes.object.isRequired,
		record: PropTypes.object,
		title: PropTypes.string.isRequired,
		url: PropTypes.string
	};
	
	return DetailsComponent;
}

export default detailsComponent;