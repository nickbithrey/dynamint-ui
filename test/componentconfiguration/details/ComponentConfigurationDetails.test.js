import React from 'react';
import { shallow } from 'enzyme';

import Details from '~/componentconfiguration/details/ComponentConfigurationDetails';
import { Selection as SelDetails } from 'office-ui-fabric-react/lib/DetailsList';

describe('<ComponentConfigurationDetails /> rendering', () => {

	let props = {
		load: jest.fn(() => []),
		record: {
			reference: 'reference',
			description: 'description',
			componentType: 'type',
			attributes: [
				{
					name: 'name',
					type: 'java.lang.Long'
				}
			],
		},
		fieldUpdate: jest.fn(() => []),
		update: jest.fn(() => []),
		location: {
			pathname: 'path',
			state: {
				uri: 'uri',
				isCreate: false
			}
		}
	};
	
	let wrapper;
	
	beforeEach(() => {
		wrapper = shallow( <Details {...props}/> );
	})
	
	it('matches snapshot', () => {
		expect(wrapper).toMatchSnapshot();
	});
	
	it('has all fields using fieldUpdate for update method', () => {
		wrapper.find('Field').forEach(f => {
			expect(f.prop('update')).toBe(props.fieldUpdate);
		});
		wrapper.find('ListField').forEach(f => {
			expect(f.prop('update')).toBe(props.fieldUpdate);
		});
	});
});