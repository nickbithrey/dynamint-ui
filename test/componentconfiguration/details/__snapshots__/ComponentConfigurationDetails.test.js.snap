// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ComponentConfigurationDetails /> rendering matches snapshot 1`] = `
<form>
  <Rowise
    tokens={
      Object {
        "childrenGap": 5,
      }
    }
  >
    <Block>
      <Field
        label="Reference"
        name="reference"
        readOnly={true}
        type="text"
        update={[MockFunction]}
        value="reference"
      />
      <Field
        label="Description"
        name="description"
        readOnly={true}
        type="text"
        update={[MockFunction]}
        value="description"
      />
      <Field
        label="Type"
        name="componentType"
        readOnly={true}
        type="text"
        update={[MockFunction]}
        value="type"
      />
      <ListField
        columns={
          Array [
            Object {
              "fieldName": "name",
              "key": "name",
              "maxWidth": 150,
              "minWidth": 50,
              "name": "Name",
            },
            Object {
              "fieldName": "type",
              "key": "type",
              "maxWidth": 50,
              "name": "Type",
            },
            Object {
              "fieldName": "required",
              "key": "required",
              "maxWidth": 10,
              "name": "Required",
              "type": "boolean",
            },
            Object {
              "fieldName": "pathParameter",
              "key": "pathParameter",
              "maxWidth": 10,
              "name": "Path Parameter",
              "type": "boolean",
            },
            Object {
              "fieldName": "defaultValue",
              "key": "defaultValue",
              "maxWidth": 150,
              "name": "Default",
            },
          ]
        }
        label="Attributes"
        name="attributes"
        newElement={
          Object {
            "componentConfiguration": "uri",
          }
        }
        readOnly={true}
        tag={[Function]}
        update={[MockFunction]}
        value={
          Array [
            Object {
              "name": "name",
              "type": "java.lang.Long",
            },
          ]
        }
      />
    </Block>
    <Block>
      <Button
        onClick={[MockFunction]}
        text="Save"
      />
    </Block>
  </Rowise>
</form>
`;
